### PR TITLE
hw5 done

### PR DESCRIPTION
**Pull request by jGrzyb**

### CODE REVIEW

File: Hw5/Task1/SparseMatrix.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            if(i < 0 || i >= Height || j < 0 || j >= Width) {
+                throw new IndexOutOfRangeException();
+            }
+            if(value != 0) {
+                nonZero[toFlattenIndex(i, j)] = value;
+            }
+        }
+    }
+
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:58]: Code style, Capital letter as first in method name. Please change in all places in code

File: Hw5/Task1/SparseMatrix.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
+using System.Collections;
+using System.Text;
+
+public class SparseMatrix : IEnumerable<long>
+{
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:5]: 1. Maybe not effective from performance point of view. You save memory, but will use ToFlattenIndex() each time you access  data. Better to use Dictionary<int, int, long> nonZero, more simple logics and no conversions from int to (int, int) and vice versa.
2. Consider situation -> Height = Int.MaxValue, Wedth int.MaxValue; Your int  FlattenIndex will be overflowed in checked context, or, worse, unexpected behavior in unchecked context  (which is used as default). Please fix, at least use  Dictionary<long, long>

File: Hw5/Task1/SparseMatrix.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    private (int a, int b) toIndexes(int flattenIndex)
+    {
+        return (flattenIndex / Width, flattenIndex % Width);
+    }
+    private int toFlattenIndex(int a, int b)
+    {
+        return a * Width + b;
+    }
+    public long[,] toMatrix()
+    {
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:68]: Bad idea, sparse matrix just to save memory, but here you allocate and fill matrix of size Height * Width, this fact ruined the idea of sparse matrix

File: Hw5/Task1/SparseMatrix.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            for(int j=0; j<Width; j++)
+            {
+                sb.Append(matrix[i, j].ToString().PadLeft(padding));
+            }
+            sb.AppendLine();
+        }
+        return sb.ToString();
+    }
+
+    public IEnumerator<long> GetEnumerator()
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:94]: use yield here

File: Hw5/Task2/Book.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+using System.Reflection;
+
+public class Book
+{
+    public string Title {get;}
+    public DateTime? ReleaseDate {get;}
+    public HashSet<string> Authors = new();
+
```
Comment (IgorKheidorov) [inline Hw5/Task2/Book.cs:8]: Code style

File: Hw5/Task2/Catalog.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public IEnumerable<Book> GetBooksByAuthor(string author)
+    {
+        return dictionary.Select(x => x.Value).Where(x => x.Authors.Contains(author)).OrderBy(x => x.ReleaseDate);
+    }
+
+    public IEnumerable<(string, int)> GetAuthorsWithBookCount()
+    {
+        return dictionary.Values
```
Comment (IgorKheidorov) [inline Hw5/Task2/Catalog.cs:47]: More simple and effective o use GroupBy (author => author ) instead of  Distinct()

File: Hw5/Task2/Catalog.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
+using System.Collections.ObjectModel;
+using System.Text.RegularExpressions;
+
+public class Catalog
+{
```
Comment (IgorKheidorov) [inline Hw5/Task2/Catalog.cs:5]: Better to implement separate class for Isbn, move all checks and conversions to it, and use     Dictionary<Isbn, Book> dictionary = new();

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please redesign the tasks 5.1 and fix comments in task 5.2

File: Hw5/Task1/SparseMatrix.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+                    this[i, j] = matrix[i, j];
+                }
+            }
+        }
+    }
+
+    public long this[int i, int j]
+    {
+        get
+        {
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:42]: Move check yo a separate private method not to duplicate code, DRY. Use it both at line  43 and 51

File: Hw5/Task1/SparseMatrix.cs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        set
+        {
+            if(i < 0 || i >= Height || j < 0 || j >= Width) 
+            {
+                throw new IndexOutOfRangeException();
+            }
+            if(value != 0) 
+            {
+                nonZero[(i, j)] = value;
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:57]: Process case if new value=0, but prev value !=0 (already in a Dictinary). Remove it from dictionary

File: Hw5/Task1/SparseMatrix.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            }
+        }
+    }
+
+    private (int a, int b) toIndexes(int flattenIndex)
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:62]: Remove unused methods

File: Hw5/Task2/Book.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
+using System.Reflection;
+
+public class Book
+{
+    public string Title { get; }
+    public DateTime? ReleaseDate { get; }
+    public HashSet<string> Authors { get; } = new();
+
+    public Book(string title, DateTime? releaseDate = null) 
+    {
```
Comment (IgorKheidorov) [inline Hw5/Task2/Book.cs:10]: Check title for null or empty

File: Hw5/Task2/Book.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        Title = title;
+        ReleaseDate = releaseDate;
+    }
+    public Book(string title, string[] authors, DateTime? releaseDate = null) 
+    {
+        Title = title;
+        ReleaseDate = releaseDate;
```
Comment (IgorKheidorov) [inline Hw5/Task2/Book.cs:17]: authors can be null

File: Hw5/Task2/Isbn.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
+using System.Text.RegularExpressions;
+
```
Comment (IgorKheidorov) [inline Hw5/Task2/Isbn.cs:2]: Cool

**Review by IgorKheidorov — CHANGES_REQUESTED** [review]
Please fix comments

File: Hw5/Task1/SparseMatrix.cs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            return get(i, j);
+        }
+        set
+        {
+            set(i, j, value);
+        }
+    }
+
+    private long get(int i, int j)
+    {
```
Comment (IgorKheidorov) [inline Hw5/Task1/SparseMatrix.cs:52]: One method for check-> both for get and set parts

Comment (jGrzyb) [reply Hw5/Task1/SparseMatrix.cs:52]: I repaired this one. Whats wrong with 'all previous'?

**Review by IgorKheidorov — APPROVED** [review]
Ok

**Pull request merged**